File: construct_erpnext/procurement_control/report/work_item_procurement_summary/work_item_procurement_summary.py
import frappe
from frappe import _
from construct_erpnext.reporting.report_utils import sum_field, summary_value, normalize_common_filters


def execute(filters=None):
	filters = normalize_common_filters(filters)
	data = get_data(filters)
	return get_columns(), data, None, None, None, False


def get_columns():
	return [
		{"label": _("Project"), "fieldname": "project", "fieldtype": "Link", "options": "Project", "width": 140},
		{"label": _("BOQ"), "fieldname": "construction_boq", "fieldtype": "Link", "options": "Construction BOQ", "width": 150},
		{"label": _("Work Item"), "fieldname": "name", "fieldtype": "Link", "options": "Construction Work Item", "width": 160},
		{"label": _("WBS"), "fieldname": "wbs_element", "fieldtype": "Link", "options": "WBS Element", "width": 140},
		{"label": _("Cost Code"), "fieldname": "cost_code", "fieldtype": "Link", "options": "Cost Code", "width": 120},
		{"label": _("Item"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 120},
		{"label": _("Description"), "fieldname": "description", "fieldtype": "Data", "width": 220},
		{"label": _("Planned Qty"), "fieldname": "planned_quantity", "fieldtype": "Float", "width": 110},
		{"label": _("Requested Qty"), "fieldname": "requested_qty", "fieldtype": "Float", "width": 110},
		{"label": _("Ordered Qty"), "fieldname": "ordered_qty", "fieldtype": "Float", "width": 110},
		{"label": _("Received Qty"), "fieldname": "received_qty", "fieldtype": "Float", "width": 110},
		{"label": _("Invoiced Qty"), "fieldname": "invoiced_qty", "fieldtype": "Float", "width": 110},
		{"label": _("Consumed Qty"), "fieldname": "consumed_qty", "fieldtype": "Float", "width": 110},
		{"label": _("Remaining Qty"), "fieldname": "remaining_qty", "fieldtype": "Float", "width": 110},
		{"label": _("Planned Amount"), "fieldname": "planned_amount", "fieldtype": "Currency", "width": 130},
		{"label": _("Committed Amount"), "fieldname": "committed_amount", "fieldtype": "Currency", "width": 140},
		{"label": _("Invoiced Amount"), "fieldname": "invoiced_amount", "fieldtype": "Currency", "width": 130},
		{"label": _("Consumed Amount"), "fieldname": "consumed_amount", "fieldtype": "Currency", "width": 140},
		{"label": _("Procurement Status"), "fieldname": "procurement_status", "fieldtype": "Data", "width": 150},
	]


def get_data(filters):
	conditions = ["disabled = 0"]
	values = {}
	for fieldname in ("project", "construction_boq", "item_category", "procurement_status"):
		if filters.get(fieldname):
			conditions.append(f"{fieldname} = %({fieldname})s")
			values[fieldname] = filters.get(fieldname)
	if filters.get("construction_work_item"):
		conditions.append("name = %(construction_work_item)s")
		values["construction_work_item"] = filters.get("construction_work_item")

	return frappe.db.sql(
		"""
		SELECT
			project, construction_boq, name, wbs_element, cost_code, item_code, description,
			planned_quantity, requested_qty, ordered_qty, received_qty, invoiced_qty,
			consumed_qty, remaining_qty, planned_amount, committed_amount, invoiced_amount,
			consumed_amount, procurement_status
		FROM `tabConstruction Work Item`
		WHERE {conditions}
		ORDER BY project, construction_boq, name
		""".format(conditions=" AND ".join(conditions)),
		values,
		as_dict=True,
	)


def get_report_summary(data):
	return [
		summary_value("Requested Qty", sum_field(data, "requested_qty"), "Float", "Blue"),
		summary_value("Ordered Qty", sum_field(data, "ordered_qty"), "Float", "Blue"),
		summary_value("Received Qty", sum_field(data, "received_qty"), "Float", "Green"),
		summary_value("Invoiced Qty", sum_field(data, "invoiced_qty"), "Float", "Orange"),
		summary_value("Consumed Qty", sum_field(data, "consumed_qty"), "Float", "Blue"),
		summary_value("Remaining Qty", sum_field(data, "remaining_qty"), "Float", "Grey"),
	]

File: construct_erpnext/procurement_control/report/work_item_procurement_summary/work_item_procurement_summary.js
function construct_erpnext_report_badge(value) {
	if (window.construct_erpnext_report_badge) {
		return window.construct_erpnext_report_badge(value);
	}
	return value || "";
}

frappe.query_reports["Work Item Procurement Summary"] = {
	filters: [
		{ fieldname: "project", label: __("Project"), fieldtype: "Link", options: "Project" },
		{ fieldname: "construction_boq", label: __("Construction BOQ"), fieldtype: "Link", options: "Construction BOQ" },
		{ fieldname: "construction_work_item", label: __("Construction Work Item"), fieldtype: "Link", options: "Construction Work Item" },
		{ fieldname: "item_category", label: __("Item Category"), fieldtype: "Select", options: "\nMaterial\nLabor\nEquipment\nSubcontract\nOverhead\nContingency\nOther" },
		{ fieldname: "procurement_status", label: __("Procurement Status"), fieldtype: "Select", options: "\nNot Requested\nPartially Requested\nFully Requested\nPartially Ordered\nFully Ordered\nPartially Received\nFully Received\nPartially Invoiced\nFully Invoiced\nOver Requested\nOver Ordered\nOver Received" },
	],
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "procurement_status") {
			return construct_erpnext_report_badge(data.procurement_status);
		}
		return value;
	},
};
